lib/scholar/metrics.ex

```diff
@@ -481,6 +481,191 @@ defmodule Scholar.Metrics do
+      iex> Scholar.Metrics.auc(fpr, tpr)
+      #Nx.Tensor<
+        f32
+        0.75
+      >
+  """
+  defn auc(x, y) do
+    check_shape(x, y)
+    dx = x[[1..-1//1]] - x[[0..-2//1]]
+    # direction = compute_direction(dx)
+
+    # 0 means x is neither increasing nor decreasing -> error
+    direction =
+      cond do
+        Nx.all(dx <= 0) -> -1
+        Nx.all(dx >= 0) -> 1
+        true -> 0
+      end
+
+    direction * trapz1d(y, x, use_x?: true)
+  end
+
+  # TODO add support for multi-dimensional x and y and move to Nx
+  defnp trapz1d(y, x, opts \\ []) do
+    opts = keyword!(opts, use_x?: false, dx: 1.0, axis: -1)
+
+    d =
+      cond do
+        opts[:use_x?] == false -> opts[:dx]
+        opts[:use_x?] == true -> x[[1..-1//1]] - x[[0..-2//1]]
+      end
+
+    Nx.sum(d * (y[[0..-2//1]] + y[[1..-1//1]]) / 2.0)
+  end
+
+  @doc ~S"""
+  It's a helper function for `Scholar.Metrics.roc_curve` and `Scholar.Metrics.roc_auc_score` functions.
+  You should call it and use as follows:
+
+    `iex> distinct_value_indices = Scholar.Metrics.calculate_distinct_value_indices(scores)`
+    `iex> {fpr, tpr, thresholds} = Scholar.Metrics.roc_curve(y_true, scores, distinct_value_indices, weights)`
+  """
+  def calculate_distinct_value_indices(y_score) do
+    desc_score_indices = Nx.argsort(y_score, direction: :desc)
+    y_score = Nx.take_along_axis(y_score, desc_score_indices)
+
+    distinct_value_indices_mask =
+      Nx.subtract(y_score[[1..-1//1]], y_score[[0..-2//1]]) |> Nx.not_equal(0)
+
+    Nx.iota({Nx.size(y_score) - 1})
+    |> Nx.add(1)
+    |> Nx.multiply(distinct_value_indices_mask)
+    |> Nx.to_flat_list()
+    |> Enum.filter(fn x -> x != 0 end)
+    |> Nx.tensor()
+    |> Nx.subtract(1)
+  end
+
+  defnp binary_clf_curve(y_true, y_score, distinct_value_indices, sample_weights) do
+    check_shape(y_true, y_score)
+
+    desc_score_indices = Nx.argsort(y_score, direction: :desc)
+    y_score = Nx.take_along_axis(y_score, desc_score_indices)
+    y_true = Nx.take_along_axis(y_true, desc_score_indices)
+
+    weight = Nx.take_along_axis(sample_weights, desc_score_indices)
+
+    threshold_idxs =
+      Nx.concatenate([distinct_value_indices, Nx.new_axis(Nx.size(y_true) - 1, -1)], axis: 0)
+
+    tps = Nx.take(Nx.cumulative_sum(y_true * weight, axis: 0), threshold_idxs)
+
+    fps = Nx.take(Nx.cumulative_sum((1 - y_true) * weight, axis: 0), threshold_idxs)
+
+    {fps, tps, Nx.take(y_score, threshold_idxs)}
+  end
+
+  # TODO implement :drop_intermediate option when function Nx.diff is implemented
+  @doc ~S"""
+  Compute Receiver operating characteristic (ROC).
+
+  Note: this implementation is restricted to the binary classification task.
+
+  ## Examples
+
+      iex> y_true = Nx.tensor([0, 0, 1, 1])
+      iex> scores = Nx.tensor([0.1, 0.4, 0.35, 0.8])
+      iex> distinct_value_indices = Scholar.Metrics.calculate_distinct_value_indices(scores)
```

Comment:
Ideally we would figure out how to implement this with `defn`, I can play with this later. :)

Reply:
This is a more general question in Nx. If you can solve this somehow, maybe the solution can be used to internally handle dynamic shapes.

Reply:
We may be able to solve this in an algorithmic way, keeping the shape full, and removing the zeros on the computation itself.

Reply:
I play with this a little and the only idea I had so far was to return the value of the previous index (instead of discarding them), but I think doing this would be incorrect on the trapeizodal function.

Reply:
Yeah, the only way we can deal with the dynamic shapes currently is to preserve the original shapes and fill the rest of the indices with some "filling'. But in 95% of cases, this filling is unwanted so we need either way a filtering function at the end of the pipeline